versiondb/tmdb/store.go

```diff
@@ -0,0 +1,217 @@
+
+import (
+	"bytes"
+	"errors"
+
+	"github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	gogotypes "github.com/cosmos/gogoproto/types"
+	"github.com/crypto-org-chain/cronos/versiondb"
+	dbm "github.com/tendermint/tm-db"
+)
+
+const latestVersionKey = "s/latest"
+
+var _ versiondb.VersionStore = (*Store)(nil)
+
+// Store implements `VersionStore`.
+type Store struct {
+	// latest key-value pairs
+	plainDB dbm.DB
+	// history bitmap index of keys
+	historyDB dbm.DB
+	// changesets of each blocks
+	changesetDB dbm.DB
+}
+
+func NewStore(plainDB, historyDB, changesetDB dbm.DB) *Store {
+	return &Store{plainDB, historyDB, changesetDB}
+}
+
+// PutAtVersion implements VersionStore interface
+// TODO reduce allocation within iterations.
+func (s *Store) PutAtVersion(version int64, changeSet []types.StoreKVPair) error {
+	plainBatch := s.plainDB.NewBatch()
+	defer plainBatch.Close()
+	historyBatch := s.historyDB.NewBatch()
+	defer historyBatch.Close()
+	changesetBatch := s.changesetDB.NewBatch()
+	defer changesetBatch.Close()
+
+	for _, pair := range changeSet {
+		key := prependStoreKey(pair.StoreKey, pair.Key)
+
+		if version == 0 {
+			// genesis state is written into plain state directly
+			if pair.Delete {
+				return errors.New("can't delete at genesis")
+			} else {
+				if err := plainBatch.Set(key, pair.Value); err != nil {
+					return err
+				}
+			}
+			continue
+		}
+
+		original, err := s.plainDB.Get(key)
+		if err != nil {
+			return err
+		}
+		if bytes.Equal(original, pair.Value) {
+			// do nothing if the value is not changed
+			continue
+		}
+
+		// write history index
+		if err := WriteHistoryIndex(s.historyDB, historyBatch, key, uint64(version)); err != nil {
+			return err
+		}
+
+		// write the old value to changeset
+		if len(original) > 0 {
+			changesetKey := append(sdk.Uint64ToBigEndian(uint64(version)), key...)
+			if err := changesetBatch.Set(changesetKey, original); err != nil {
+				return err
+			}
+		}
+
+		// write the new value to plain state
+		if pair.Delete {
+			if err := plainBatch.Delete(key); err != nil {
+				return err
+			}
+		} else {
+			if err := plainBatch.Set(key, pair.Value); err != nil {
+				return err
+			}
+		}
+	}
+
+	// write latest version to plain state
+	if err := s.setLatestVersion(plainBatch, version); err != nil {
+		return err
+	}
+
+	if err := changesetBatch.WriteSync(); err != nil {
+		return err
+	}
+	if err := historyBatch.WriteSync(); err != nil {
+		return err
+	}
+	return plainBatch.WriteSync()
+}
+
+// GetAtVersion implements VersionStore interface
+func (s *Store) GetAtVersion(storeKey string, key []byte, version *int64) ([]byte, error) {
+	rawKey := prependStoreKey(storeKey, key)
+	if version == nil {
+		return s.plainDB.Get(rawKey)
+	}
+
+	height := *version
+
+	// optimize for latest version
+	latest, err := s.GetLatestVersion()
+	if err != nil {
+		return nil, err
+	}
+	if latest == height {
+		return s.plainDB.Get(rawKey)
+	}
+
```

Comment:
maybe return nil or error here if `height` > `latest` ?

Reply:
>= Sounds good

Reply:
I added a check, so it'll fail if height is in the future.